### PR TITLE
[Fiber] Clear existing text content before inserting children

### PR DESCRIPTION
**Pull request by acdlite**
Fixes a bug when updating from a single text child (or `dangerouslySetInnerHTML`) to regular children, where the previous text content never gets deleted.

Based on https://github.com/facebook/react/pull/8304 ([compare view](https://github.com/acdlite/react/compare/fibererroreffect...acdlite:fiberresettextcontent)).

### CODE REVIEW

**Comment by acdlite** [comment]
Here's a description of what I've added:

- In reconciliation phase, check if the host component is going from text children to normal children. If so, schedule a `ContentReset` effect. The logic for checking if the host component should render with a text child is now part of the host config. We can't just check for text and number children because of special cases like `__dangerouslySetInnerHTML`.
- Whenever a node is being inserted, check if the parent has a `ContentReset` effect. If so, reset the text content of the parent and remove the flag from the effect tag. Resetting the text content of a host instance was also added to the host config.
- In the commit phase, perform any `ContentReset` effects that weren't already handled by child insertions.

File: src/renderers/dom/fiber/ReactDOMFiber.js [hunk 1]
```diff
@@ -89,6 +89,22 @@ var DOMRenderer = ReactFiberReconciler({
...
+      (
+        typeof props.dangerouslySetInnerHTML === 'object' &&
+        props.dangerouslySetInnerHTML !== null &&
+        typeof props.dangerouslySetInnerHTML.__html === 'string'
+      )
+    );
+  },
+
+  resetTextContent(domElement : Instance) : void {
+    domElement.textContent = '';
```
Comment (syranide) [inline src/renderers/dom/fiber/ReactDOMFiber.js]: If we're concerned with performance then IIRC setting textContent or innerHTML is unintuitively significantly slower than simply clearing the children one by one.

**Comment by joh1234** [comment]
Update README
https://goo.gl/JBljdR

or 
http://bit.ly/2dK6Ojk

**Comment by sebmarkbage** [comment]
Ok, accepting this but not the things it build on top of. So this will either have to wait or rebase on top of master.

File: src/renderers/dom/fiber/ReactDOMFiber.js (hunk 1 above)
Comment (acdlite) [inline src/renderers/dom/fiber/ReactDOMFiber.js]: Oh right that's true

File: src/renderers/shared/fiber/ReactFiberBeginWork.js [hunk 2]
```diff
@@ -210,14 +211,27 @@ module.exports = function<T, P, I, TI, C>(
   }
 
   function updateHostComponent(current, workInProgress) {
+    const nextProps = workInProgress.pendingProps;
+    const prevProps = current ? current.memoizedProps : null;
     let nextChildren = workInProgress.pendingProps.children;
-    if (typeof nextChildren === 'string' || typeof nextChildren === 'number') {
+    const isDirectTextChild = config.shouldSetTextContent(nextProps);
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: Nit: This should've been cached in a local variable above instead of using the object form.

**Comment by arcanis** [comment]
Just a small feedback after implementing `shouldSetTextContent` in a custom renderer: it could be useful if the hook could also get the type of the parent instance. It would enable this kind of logic, where we want to ensure that texts are wrapped into a specific type, but also want to allow using this type directly:

```js
createInstance(type, props) {
    return new (components.get(type))(props);
}

createTextInstance(text) {
    return new (components.get('text'))({ text });
}

shouldSetTextContent(type, props) {
    return type === 'text';
}
```

**Pull request merged**